Design of data preprocessing in AutoML.md

```diff
@@ -0,0 +1,93 @@
+
+#### Cons:
+- wrapping original main model into PipelineModel will hide its functionality. It can be mitigated by introducing ProxyModel interface with `getMainModel` method that will provide main model to the user. Though, for the clients (theoretically) PipelineModel could be flattenned automatically. Complexity of doing this  depends on how flexible we are with serialisation at rest api layer.
+
+![PipelineModel approach](pipelinemodel.png)
+
+(2) *Integration preprocessing functionality into a Model class*
+
+We can probably have two sub-approaches here:
+- when we have PipelineModelBuilder and during training it collapses into main model without any wrapper around it
+- when we extend functionality of ModelBuilder by adding ability to have references to proprocessing models
+
+Not sure which approach is better but probably we can avoid introduction of the PipelineModelBuilder class. But it will increase number of responsibilities of the ModelBuilder class.
+
+#### Pros:
+- We do not hide any functionality behind PipelineModel wrapper
+- All models could be equally seen as pipelines ( which could also be a cons.)
+
+#### Cons:
+- conceptually abusing notion of the Model. Model ( what is also known as Transformer in other frameworks) is supposed to be only trained on data and after that be able to transform similar data. Nothing about data preprocessing is expected here. See Concerns section for more details.
+- new functionality will be exposed not only to AutoML scope. All algos will be "aware" about data preprocessing from now on. With first approach we would just add notion of ability to be a Proxy model and not all the models will have to realise this interface. Though, absense of the preprocessing models could be seen as "no-impact" on Model, still from code perspective there will a bigger impact. Hopefully preprocessing functionality will be generic enough without depenedencies on a particular types of the models.
+- increased number of responsibilities of the ModelBuilder and Model, which could be otherwise be located in other classes.
+
+
+![Integration preprocessing functionality into a Model class](no_pipelinemodel.png)
+
+
+#### General concerns:
+Wanted one more time to draw our attention to the fact how we handle cornerstone abstractions in our project. And let's say we are not going to use Pipelines.
+
+For comparison I will be using SparkML framework:
+
+![SparkML vs H2O.ai](sparkml_vs_h2oai.png)
+
+ref. https://www.youtube.com/watch?v=jEyahxFp3ak
+
+Model is being overflowed with responsibilities imho. Transformation logic is being spread over many places in different forms.
+
+### Summary
+- we tend to keep adding functionality into Model class. We might be going to add pipeline related aspect into Model
```

Comment:
All encodings are handled on the Model level, from this point of view it is internally consistent.

Reply:
@michalkurka it would be consistent internally, true. But I want to highlight here that I separate Transformation aspect from Pipelining aspect. We already have both of them hardcoded inside Model. And my opinion is that these aspects could be handled separately.